src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableOutputConnection.cs

```diff
@@ -0,0 +1,362 @@
+                }
+                finally
+                {
+                    _sendGuard.Exit();
+                }
+            }
+        }
+
+        private void OnTransferComplete()
+        {
+            Strategy.DequeuePending();
+
+            if (Strategy.DoneTransmitting)
+                Terminate();
+        }
+
+        public void ProcessTransferred(long transferred, SequenceRangeCollection ranges, int quotaRemaining)
+        {
+            if (transferred < 0)
+            {
+                throw Fx.AssertAndThrow("Argument transferred must be a valid sequence number or 0 for protocol messages.");
+            }
+
+            bool invalidAck;
+
+            // ignored, TransmissionStrategy is being used to keep track of what must be re-sent.
+            // In the Request-Reply case this state may not align with acks.
+            bool inconsistentAck;
+
+            Strategy.ProcessAcknowledgement(ranges, out invalidAck, out inconsistentAck);
+            invalidAck = (invalidAck || ((transferred != 0) && !ranges.Contains(transferred)));
+
+            if (!invalidAck)
+            {
+                if ((transferred > 0) && Strategy.ProcessTransferred(transferred, quotaRemaining))
+                {
+                    ActionItem.Schedule(s_sendRetries, this);
+                }
+                else
+                {
+                    OnTransferComplete();
+                }
+            }
+            else
+            {
+                WsrmFault fault = new InvalidAcknowledgementFault(_id, ranges);
+                RaiseFault(fault.CreateException(), fault);
+            }
+        }
+
+        public void ProcessTransferred(SequenceRangeCollection ranges, int quotaRemaining)
+        {
+            bool invalidAck;
+            bool inconsistentAck;
+
+            Strategy.ProcessAcknowledgement(ranges, out invalidAck, out inconsistentAck);
+
+            if (!invalidAck && !inconsistentAck)
+            {
+                if (Strategy.ProcessTransferred(ranges, quotaRemaining))
+                {
+                    ActionItem.Schedule(s_sendRetries, this);
+                }
+                else
+                {
+                    OnTransferComplete();
+                }
+            }
+            else
+            {
+                WsrmFault fault = new InvalidAcknowledgementFault(_id, ranges);
+                RaiseFault(fault.CreateException(), fault);
+            }
+        }
+
+        private void RaiseFault(Exception faultException, WsrmFault fault)
+        {
+            ComponentFaultedHandler handler = Faulted;
+
+            if (handler != null)
+                handler(faultException, fault);
+        }
+
+        private void RaiseOnException(Exception exception)
+        {
+            ComponentExceptionHandler handler = OnException;
+
+            if (handler != null)
+                handler(exception);
+        }
+
+        private async void SendRetries()
```

Comment:
We generally don't use `async void`; we just use `async Task` and swallow the result. `async void` does one or two things you probably don't care about, to support its use in UI scenarios like `button.OnClick`. It's not a big deal though.

Reply:
I've fixed most of these and they'll be in the next push. I had to leave a couple as they actually are set on an EventHandler but all the ones called as delegates or fired from a timer have been converted to async. I had to modify IOThreadTimer and ActionItem.Schedule to support passing an async Func and create a TaskScheduler which posts continuations to our IOThreadScheduler as the synchronous versions of these methods ran exclusively on IO Threads.